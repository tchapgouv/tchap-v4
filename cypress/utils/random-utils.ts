export default class RandomUtils {
    /**
     * Generate a random string to a length of max 10 char
     * @param length < 10
     * @returns random string
     */
    public static generateRandom(length: number): string {
        return (Math.random() + 1).toString(36).substring(2, 2 + length);
    }
}
