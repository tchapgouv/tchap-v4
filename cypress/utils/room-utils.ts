import Chainable = Cypress.Chainable;
import TchapCreateRoom from "../../src/lib/createTchapRoom";
import { TchapRoomType } from "../../src/@types/tchap";
export default class RoomUtils {
    public static createPublicRoom(roomName: string): Chainable<string> {
        return cy.createRoom(TchapCreateRoom.roomCreateOptions(roomName, TchapRoomType.Forum, false).createOpts);
    }
    public static createPrivateRoom(roomName: string): Chainable<string> {
        return cy.createRoom(TchapCreateRoom.roomCreateOptions(roomName, TchapRoomType.Private, false).createOpts);
    }
    public static createPrivateWithExternalRoom(roomName: string): Chainable<string> {
        return cy.createRoom(TchapCreateRoom.roomCreateOptions(roomName, TchapRoomType.External, false).createOpts);
    }
    public static openRoomAccessSettings(roomName: string): Chainable<JQuery<HTMLElement>> {
        //open room
        cy.get('[aria-label="' + roomName + '"]').click();
        cy.get(".mx_RoomHeader_chevron").click();
        cy.get('[aria-label="ParamÃ¨tres"] > .mx_IconizedContextMenu_label').click();
        return cy.get('[data-testid="settings-tab-ROOM_SECURITY_TAB"] > .mx_TabbedView_tabLabel_text').click();
    }
}
