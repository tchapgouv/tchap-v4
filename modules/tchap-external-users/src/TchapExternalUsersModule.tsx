/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
