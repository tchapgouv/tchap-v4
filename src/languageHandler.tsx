/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
