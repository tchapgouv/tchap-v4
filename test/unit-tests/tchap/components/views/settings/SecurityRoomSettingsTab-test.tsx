/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { render, screen, waitFor } from "jest-matrix-react";
import { EventType, GuestAccess, HistoryVisibility, JoinRule, MatrixEvent, Room } from "matrix-js-sdk/src/matrix";
import { mocked } from "jest-mock";

import SecurityRoomSettingsTab from "~tchap-web/src/components/views/settings/tabs/room/SecurityRoomSettingsTab";
import MatrixClientContext from "~tchap-web/src/contexts/MatrixClientContext";
import SettingsStore from "~tchap-web/src/settings/SettingsStore";
import { clearAllModals, stubClient } from "~tchap-web/test/test-utils";
import { filterBoolean } from "~tchap-web/src/utils/arrays";

describe("<SecurityRoomSettingsTab />", () => {
    const userId = "@alice:server.org";
    const client = mocked(stubClient());
    const roomId = "!room:server.org";

    const getComponent = (room: Room, closeSettingsFn = jest.fn()) =>
        render(<SecurityRoomSettingsTab room={room} closeSettingsFn={closeSettingsFn} />, {
            wrapper: ({ children }) => (
                <MatrixClientContext.Provider value={client}>{children}</MatrixClientContext.Provider>
            ),
        });

    const setRoomStateEvents = (
        room: Room,
        joinRule?: JoinRule,
        guestAccess?: GuestAccess,
        history?: HistoryVisibility,
    ): void => {
        const events = filterBoolean<MatrixEvent>([
            new MatrixEvent({
                type: EventType.RoomCreate,
                content: { version: "test" },
                sender: userId,
                state_key: "",
                room_id: room.roomId,
            }),
            guestAccess &&
                new MatrixEvent({
                    type: EventType.RoomGuestAccess,
                    content: { guest_access: guestAccess },
                    sender: userId,
                    state_key: "",
                    room_id: room.roomId,
                }),
            history &&
                new MatrixEvent({
                    type: EventType.RoomHistoryVisibility,
                    content: { history_visibility: history },
                    sender: userId,
                    state_key: "",
                    room_id: room.roomId,
                }),
            joinRule &&
                new MatrixEvent({
                    type: EventType.RoomJoinRules,
                    content: { join_rule: joinRule },
                    sender: userId,
                    state_key: "",
                    room_id: room.roomId,
                }),
        ]);

        room.currentState.setStateEvents(events);
    };

    beforeEach(async () => {
        client.sendStateEvent.mockReset().mockResolvedValue({ event_id: "test" });
        jest.spyOn(client.getCrypto()!, "isEncryptionEnabledInRoom").mockResolvedValue(false);
        client.getClientWellKnown.mockReturnValue(undefined);
        jest.spyOn(SettingsStore, "getValue").mockRestore();

        await clearAllModals();
    });

    describe("history visibility", () => {
        it("does not render shared readable option when room is encrypted", async () => {
            const room = new Room(roomId, client, userId);
            jest.spyOn(client.getCrypto()!, "isEncryptionEnabledInRoom").mockResolvedValue(true);
            setRoomStateEvents(room);

            getComponent(room);

            await waitFor(() => expect(screen.queryByDisplayValue(HistoryVisibility.Shared)).not.toBeInTheDocument());
        });

        it("uses shared as default history visibility when no state event found", () => {
            const room = new Room(roomId, client, userId);
            setRoomStateEvents(room);

            getComponent(room);

            expect(screen.getByText("Who can read history?").parentElement).toMatchSnapshot();
            expect(screen.getByDisplayValue(HistoryVisibility.Shared)).toBeChecked();
        });
    });
});
