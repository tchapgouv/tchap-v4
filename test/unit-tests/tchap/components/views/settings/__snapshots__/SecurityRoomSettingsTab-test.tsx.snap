// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SecurityRoomSettingsTab /> history visibility uses shared as default history visibility when no state event found 1`] = `
<fieldset
  class="mx_SettingsFieldset"
>
  <legend
    class="mx_SettingsFieldset_legend"
  >
    Who can read history?
  </legend>
  <div
    class="mx_SettingsFieldset_description"
  >
    <div
      class="mx_SettingsSubsection_text"
    >
      Changes to who can read history will only apply to future messages in this room. The visibility of existing history will be unchanged.
    </div>
  </div>
  <div
    class="mx_SettingsFieldset_content"
  >
    <label
      class="mx_StyledRadioButton mx_StyledRadioButton_enabled"
    >
      <input
        id="historyVis-world_readable"
        name="historyVis"
        type="radio"
        value="world_readable"
      />
      <div>
        <div />
      </div>
      <div
        class="mx_StyledRadioButton_content"
      >
        Anyone
      </div>
      <div
        class="mx_StyledRadioButton_spacer"
      />
    </label>
    <label
      class="mx_StyledRadioButton mx_StyledRadioButton_enabled mx_StyledRadioButton_checked"
    >
      <input
        checked=""
        id="historyVis-shared"
        name="historyVis"
        type="radio"
        value="shared"
      />
      <div>
        <div />
      </div>
      <div
        class="mx_StyledRadioButton_content"
      >
        Members only (since the point in time of selecting this option)
      </div>
      <div
        class="mx_StyledRadioButton_spacer"
      />
    </label>
    <label
      class="mx_StyledRadioButton mx_StyledRadioButton_enabled"
    >
      <input
        id="historyVis-invited"
        name="historyVis"
        type="radio"
        value="invited"
      />
      <div>
        <div />
      </div>
      <div
        class="mx_StyledRadioButton_content"
      >
        Members only (since they were invited)
      </div>
      <div
        class="mx_StyledRadioButton_spacer"
      />
    </label>
    <label
      class="mx_StyledRadioButton mx_StyledRadioButton_enabled"
    >
      <input
        id="historyVis-joined"
        name="historyVis"
        type="radio"
        value="joined"
      />
      <div>
        <div />
      </div>
      <div
        class="mx_StyledRadioButton_content"
      >
        Members only (since they joined)
      </div>
      <div
        class="mx_StyledRadioButton_spacer"
      />
    </label>
  </div>
</fieldset>
`;
